Stripe/STPCardValidator.swift

```diff
@@ -0,0 +1,445 @@
+        }
+      }
+    default:
+      return .invalid
+    }
+  }
+
+  class func validationState(
+    forCard card: STPCardParams, inCurrentYear currentYear: Int, currentMonth: Int
+  ) -> STPCardValidationState {
+    let numberValidation = self.validationState(
+      forNumber: card.number ?? "", validatingCardBrand: true)
+    let expMonthString = String(format: "%02lu", UInt(card.expMonth))
+    let expMonthValidation = self.validationState(forExpirationMonth: expMonthString)
+    let expYearString = String(format: "%02lu", UInt(card.expYear) % 100)
+    let expYearValidation = self.validationState(
+      forExpirationYear: expYearString,
+      inMonth: expMonthString,
+      inCurrentYear: currentYear,
+      currentMonth: currentMonth)
+    let brand = self.brand(forNumber: card.number ?? "")
+    let cvcValidation = self.validationState(forCVC: card.cvc ?? "", cardBrand: brand)
+
+    let states = [
+      NSNumber(value: numberValidation.rawValue),
+      NSNumber(value: expMonthValidation.rawValue),
+      NSNumber(value: expYearValidation.rawValue),
+      NSNumber(value: cvcValidation.rawValue),
+    ]
+    var incomplete = false
+    for boxedState in states {
+      let state = STPCardValidationState(rawValue: boxedState.intValue)
+      if state == .invalid {
+        return state!
+      } else if state == .incomplete {
+        incomplete = true
+      }
+    }
+    return incomplete ? .incomplete : .valid
+  }
+
+  class func minCVCLength() -> Int {
+    return 3
+  }
+
+  class func possibleBrands(forNumber cardNumber: String) -> Set<STPCardBrand> {
+    let binRanges = STPBINRange.binRanges(forNumber: cardNumber)
+    var brands = binRanges.map(\.brand)
+    brands.removeAll { $0 == .unknown }
+    return Set(brands)
+  }
+
+  class func currentYear() -> Int {
+    let calendar = Calendar(identifier: .gregorian)
+    return (calendar.component(.year, from: Date())) % 100
+  }
+
+  class func currentMonth() -> Int {
+    let calendar = Calendar(identifier: .gregorian)
+    return calendar.component(.month, from: Date())
+  }
+}
+
+extension STPCardValidator {
+  class func cardNumberFormat(for brand: STPCardBrand) -> [NSNumber] {
+    switch brand {
+    case .amex:
+      return [NSNumber(value: 4), NSNumber(value: 6), NSNumber(value: 5)]
+    default:
+      return [NSNumber(value: 4), NSNumber(value: 4), NSNumber(value: 4), NSNumber(value: 4)]
+    }
+  }
+
+  class func cardNumberFormat(forCardNumber cardNumber: String) -> [NSNumber] {
+    let binRange = STPBINRange.mostSpecificBINRange(forNumber: cardNumber)
+    if binRange.brand == .dinersClub && binRange.length == 14 {
+      return [NSNumber(value: 4), NSNumber(value: 6), NSNumber(value: 4)]
+    }
+
+    return self.cardNumberFormat(for: binRange.brand)
+  }
+
+  class func stringIsValidLuhn(_ number: String) -> Bool {
+    var odd = true
+    var sum = 0
+    var digits = [AnyHashable](repeating: 0, count: number.count)
```

Comment:
this plus the append below will create an array with a bunch of prepended zeros

Reply:
what a weird way to convert `arrayWithCapacity`